reark/src/main/java/io/reark/reark/data/stores/cores/ContentProviderStoreCoreBase.java

```diff
@@ -89,38 +142,79 @@ public static Handler createHandler(@NonNull final String name) {
         return new Handler(handlerThread.getLooper());
     }
 
-    private void updateIfValueChanged(@NonNull final Pair<U, Uri> pair) {
-        final Cursor cursor = contentResolver.query(pair.second, getProjection(), null, null, null);
-        U newItem = pair.first;
-        boolean valuesEqual = false;
+    /**
+     * Implements grouping logic for batching the content provider operations. The default
+     * logic buffers the operations with debounced timer, while applying a hard limit for the
+     * number of operations. The data is serialized into a binder transaction, and an attempt
+     * to pass a too large batch of operations will result in a failed binder transaction.
+     */
+    @NonNull
+    protected Observable<List<ContentProviderOperation>> groupOperations(@NonNull final Observable<ContentProviderOperation> source) {
+        return source.publish(stream -> stream.buffer(
+                Observable.amb(
+                        stream.debounce(groupingTimeout, TimeUnit.MILLISECONDS),
+                        stream.skip(groupMaxSize - 1))
+                        .first() // Complete observable after the first reached trigger
+                        .repeatWhen(observable -> observable))); // Resubscribe immediately for the next buffer
+    }
 
-        if (cursor != null) {
-            if (cursor.moveToFirst()) {
-                U currentItem = read(cursor);
-                valuesEqual = newItem.equals(currentItem);
-
-                if (!valuesEqual) {
-                    Log.v(TAG, "Merging values at " + pair.second);
-                    newItem = mergeValues(currentItem, newItem);
-                    valuesEqual = newItem.equals(currentItem);
-                }
-            }
-            cursor.close();
-        }
+    @NonNull
+    private Observable<ContentProviderOperation> createOperation(@NonNull final U item, @NonNull final Uri uri) {
```

Comment:
I managed to refactor it to a bit shorter. It's still a bit on the long side, but hard to shorten more without e.g. moving cursor closing to helper methods, which I wouldn't want to do as it would become easier to break.